### PR TITLE
Image classification preview 2.

### LINKED ISSUE #4090: [ImageClassification Transfer Learning] - Warning mesage: "Allocation of 553190400 exceeds 10% of system memory."
Why are we getting this warning message (I get it multiple times) when training with a small set of images?

"2019-08-07 09:49:52.642800: W tensorflow/core/framework/allocator.cc:107] Allocation of 553190400 exceeds 10% of system memory."

image

### PR DESCRIPTION
**Pull request by codemzs**
This change improves the in-preview image classification API further:

1. Increases DNN training speed by 10x.
2. Reduced and constant memory footprint.
3. Simplifies the API by not requiring the user to pre-process the image.
4. Introduces callback to provide metrics during training such as accuracy, cross-entropy.
5. Improved image classification sample.

fixes #4090, #4087, #4084, #4134

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/ImageClassification/ResnetV2101TransferLearning.cs [hunk 1]
```diff
@@ -1,123 +0,0 @@
...
-            var pipeline = mlContext.Transforms.Conversion.MapValueToKey("Label")
-                .Append(mlContext.Transforms.LoadImages("ImageObject", null,
-                    "ImagePath"))
-                .Append(mlContext.Transforms.ResizeImages("Image",
-                    inputColumnName: "ImageObject", imageWidth: 299,
-                    imageHeight: 299))
-                .Append(mlContext.Transforms.ExtractPixels("Image",
-                    interleavePixelColors: true))
-                .Append(mlContext.Model.ImageClassification("Image",
-                    "Label", arch: DnnEstimator.Architecture.ResnetV2101, epoch: 4,
```
Comment (CESARDELATORRE) [inline docs/samples/Microsoft.ML.Samples/Dynamic/ImageClassification/ResnetV2101TransferLearning.cs]: Maybe add a comment around here saying that just by changing/selecting InceptionV3 here instead of ResnetV2101 you can try a different architecture/pre-trained model. :)

File: docs/samples/Microsoft.ML.Samples/Dynamic/ImageClassification/ResnetV2101TransferLearningTrainTestSplit.cs [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+            try
+            {
+
+                MLContext mlContext = new MLContext(seed: 1);
+
+                //Load all the original images info
+                IEnumerable<ImageData> images = LoadImagesFromDirectory(folder: fullImagesetFolderPath, useFolderNameasLabel: true);
+                IDataView shuffledFullImagesDataset = mlContext.Data.ShuffleRows(mlContext.Data.LoadFromEnumerable(images));
+                shuffledFullImagesDataset = mlContext.Transforms.Conversion.MapValueToKey("Label")
+                    .Fit(shuffledFullImagesDataset)
```
Comment (CESARDELATORRE) [inline docs/samples/Microsoft.ML.Samples/Dynamic/ImageClassification/ResnetV2101TransferLearningTrainTestSplit.cs]: Let's keep it like this in any case. The only minor issue I see here is that if the user wants to remove the ValidationSet for the training pipeline and wants to simplify and do a single Fit() only for training, the changes in the code are more complex because he also has to realize that he needs to change the provided schema when saving the model. But it doesn't matter, I personally think you always will want to have the  ValidationSet for the training pipeline and see how it evolves while training... :)

File: docs/samples/Microsoft.ML.Samples/Dynamic/ImageClassification/ResnetV2101TransferLearningTrainTestSplit.cs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+                            batchSize: 10,
+                            learningRate: 0.01f,
+                            metricsCallback: (metrics) => Console.WriteLine(metrics),
+                            validationSet: testDataset);
+
+
+                Console.WriteLine("*** Training the image classification model with DNN Transfer Learning on top of the selected pre-trained model/architecture ***");
+
+                // Measuring training time
+                var watch = System.Diagnostics.Stopwatch.StartNew();
```
Comment (CESARDELATORRE) [inline docs/samples/Microsoft.ML.Samples/Dynamic/ImageClassification/ResnetV2101TransferLearningTrainTestSplit.cs]: Measuring the time with watch makes the code a bit more verbose or difficult to read.
But to me it is useful, so we can keep it here, too. 

I'm removing it from the samples repo so the code is simpler, though.

File: src/Microsoft.ML.Dnn/DnnCatalog.cs [hunk 4]
```diff
@@ -85,33 +83,50 @@ public static class DnnCatalog
...
             string featuresColumnName,
             string labelColumnName,
-            string outputGraphPath = null,
             string scoreColumnName = "Score",
             string predictedLabelColumnName = "PredictedLabel",
-            string checkpointName = "_retrain_checkpoint",
-            Architecture arch = Architecture.ResnetV2101,
-            DnnFramework dnnFramework = DnnFramework.Tensorflow,
+            Architecture arch = Architecture.InceptionV3,
             int epoch = 10,
```
Comment (CESARDELATORRE) [inline src/Microsoft.ML.Dnn/DnnCatalog.cs]: Why epoch=10 and batchSize=20 as defaults?
I find that for small/medium datasets (hundres of images) which is what initially folks are going to test, these defaults are what we were using:

epoch=100 
batchSize=10

But I'm open to different/better defaults. IT is simply that I think these defaults were defined in previous versions and we might want to change it?

Of course, I know that we cannot have good defaults for any scenario. We'd need to have some "AutoML" that would change dynamically depending on the input datasets type/size and even other hyper-parameters depending on the accuracy we're getting...

File: src/Microsoft.ML.Dnn/DnnCatalog.cs [hunk 5]
```diff
@@ -85,33 +83,50 @@ public static class DnnCatalog
...
             int batchSize = 20,
-            float learningRate = 0.01f)
+            float learningRate = 0.01f,
+            ImageClassificationMetricsCallback metricsCallback = null,
+            int statisticFrequency = 1,
+            DnnFramework framework = DnnFramework.Tensorflow,
```
Comment (CESARDELATORRE) [inline src/Microsoft.ML.Dnn/DnnCatalog.cs]: I wonder if we have external hyper-parameters that only are applicable to TensorFlow but not to PyTorch?
Can you confirm that? - In any case, we'll be able to evolve it when implementing for PyTorch, let's not start building from the roof... ;) - Never mind about this comment, just an observation.

File: src/Microsoft.ML.Dnn/DnnCatalog.cs [hunk 6]
```diff
@@ -85,33 +83,50 @@ public static class DnnCatalog
...
+            string modelSavePath = null,
+            string checkpointName = "_retrain_checkpoint",
```
Comment (CESARDELATORRE) [inline src/Microsoft.ML.Dnn/DnnCatalog.cs]: I think we need to change the name of the custom model generated for the user, I mean the .PB TensorFlow model file. Currently I think we have:

- `resnet_v2_101_299.meta_retrain_checkpoint.pb`
or if using Inception:
- `InceptionV3.meta_retrain_checkpoint.pb`

When the user sees those names, it'll feel like something related to an internal process (checkpoint, etc.).
Can you change it to something like the following?

- `custom.retrained.model.based.on.resnet_v2_101_299.pb`
or if using InceptionV3:
- `custom.retrained.model.based.on.InceptionV3.pb`

I'm open to comparable names, but using "checkpoint" will feel weird for a user. Feels like it is a checkpoint but still not finalized..

File: src/Microsoft.ML.Dnn/DnnCatalog.cs [hunk 7]
```diff
@@ -85,33 +83,50 @@ public static class DnnCatalog
...
+            IDataView validationSet = null,
+            bool testOnTrainSet = true,
+            bool reuseTrainSetBottleneckCachedValues = false,
```
Comment (CESARDELATORRE) [inline src/Microsoft.ML.Dnn/DnnCatalog.cs]: Good, cahce defaults = false. 👍

File: docs/samples/Microsoft.ML.Samples/Program.cs [hunk 8]
```diff
@@ -10,7 +10,7 @@ public static class Program
 
         internal static void RunAll()
         {
-            int samples = 0;
+            /*int samples = 0;
```
Comment (CESARDELATORRE) [inline docs/samples/Microsoft.ML.Samples/Program.cs]: Why is this code commented? 
Shall we either remove it or keep it un-commented?

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 9]
```diff
@@ -0,0 +1,1224 @@
...
+        private void CheckTrainingParameters(ImageClassificationEstimator.Options options)
+        {
+            Host.CheckNonWhiteSpace(options.LabelColumn, nameof(options.LabelColumn));
+            Host.CheckNonWhiteSpace(options.TensorFlowLabel, nameof(options.TensorFlowLabel));
+
+            if (_session.graph.OperationByName(options.TensorFlowLabel) == null)
+                throw Host.ExceptParam(nameof(options.TensorFlowLabel), $"'{options.TensorFlowLabel}' does not exist in the model");
+        }
+
+        private (Tensor, Tensor) AddJpegDecoding(int height, int width, int depth)
```
Comment (CESARDELATORRE) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: We'll need to support additional image formats, this is only for JPEG..

File: docs/samples/Microsoft.ML.Samples/Microsoft.ML.Samples.csproj [hunk 10]
```diff
@@ -71,6 +71,7 @@
       <DependentUpon>LbfgsLogisticRegressionWithOptions.tt</DependentUpon>
     </None>
     <PackageReference Include="Microsoft.ML.TensorFlow.Redist" Version="0.10.0" />
+    <PackageReference Include="SharpZipLib" Version="1.2.0" />
```
[the remark this answers is not in the document]
Comment (CESARDELATORRE) [reply docs/samples/Microsoft.ML.Samples/Microsoft.ML.Samples.csproj]: It's used only in the DNN ResNet sample to unzip the sample dataset, after downloading the .zip of the dataset. IT's used by a utility class that's why you might not see it used directly.

File: test/Microsoft.ML.Tests/ImageClassification.cs [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+        public static IEnumerable<ImageData> LoadImagesFromDirectory(string folder,
+            bool useFolderNameasLabel = true)
+        {
+            var files = Directory.GetFiles(folder, "*",
+                searchOption: SearchOption.AllDirectories);
+
+            foreach (var file in files)
+            {
+                if (Path.GetExtension(file) != ".jpg")
```
[the remark this answers is not in the document]
Comment (CESARDELATORRE) [reply test/Microsoft.ML.Tests/ImageClassification.cs]: I opened and issue for that:
https://github.com/dotnet/machinelearning/issues/4152

This is Zeeshan's answer: "The current tf.net version we are using does not seem to have bindings for all image types. But the latest version seems to have we will upgrade to them latest next month"

[the remark this answers is not in the document]
Comment (Nucs) [reply test/Microsoft.ML.Tests/ImageClassification.cs]: We are planning to release the new TF.NET version in a day or two.

[the remark this answers is not in the document]
Comment (codemzs) [reply test/Microsoft.ML.Tests/ImageClassification.cs]: @eerhardt @CESARDELATORRE The above answers your question :)

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 12]
```diff
@@ -0,0 +1,1224 @@
...
+        {
+            var labelColumn = input.Schema[labelColumnName];
+            var imagePathColumn = input.Schema[imagepathColumnName];
+            Runner runner = new Runner(_session);
+            runner.AddOutputs(outputTensorName);
+
+            using (TextWriter writer = File.CreateText(cacheFilePath))
+            using (var cursor = input.GetRowCursor(input.Schema.Where(c => c.Index == labelColumn.Index || c.Index == imagePathColumn.Index)))
+            {
+                var labelGetter = cursor.GetGetter<uint>(labelColumn);
```
[the remark this answers is not in the document]
Comment (codemzs) [reply src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: spoke offline, good to support all types but for now just throw a meaningful exception. UInt32 is default for value type.

---
In reply to: [319128148](https://github.com/dotnet/machinelearning/pull/4151#discussion_r319128148) [](ancestors = 319128148)

File: test/Microsoft.ML.Tests/ImageClassification.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Oceania2018) [reply test/Microsoft.ML.Tests/ImageClassification.cs]: Correct, we have added a lot of APIs in this [commit](https://github.com/SciSharp/TensorFlow.NET/commit/9d90d74ddd85ed4e50c77463433567005355ea0f). And we can easily add more if needed.

File: docs/samples/Microsoft.ML.Samples/Microsoft.ML.Samples.csproj (hunk 10 above)
[the remark this answers is not in the document]
Comment (Oceania2018) [reply docs/samples/Microsoft.ML.Samples/Microsoft.ML.Samples.csproj]: Why do we still use `Microsoft.ML.TensorFlow.Redist` ? Shoud refer to 
 [SciSharp.TensorFlow.Redist](https://www.nuget.org/packages/SciSharp.TensorFlow.Redist/)

File: docs/samples/Microsoft.ML.Samples/Dynamic/ImageClassification/ResnetV2101TransferLearningTrainTestSplit.cs [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+            string fullImagesetFolderPath = Path.Combine(
+                imagesDownloadFolderPath, finalImagesFolderName);
+
+            try
+            {
+
+                MLContext mlContext = new MLContext(seed: 1);
+
+                //Load all the original images info
+                IEnumerable<ImageData> images = LoadImagesFromDirectory(
```
Comment (CESARDELATORRE) [inline docs/samples/Microsoft.ML.Samples/Dynamic/ImageClassification/ResnetV2101TransferLearningTrainTestSplit.cs]: LoadImagesFromDirectory() is a method that is almost always going to be used by developers.
I think we should consider to have this method provided by the framework.
This is also something being requested by customers.

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 14]
```diff
@@ -1113,5 +1205,183 @@ public void TensorFlowStringTest()
...
+
+        public static string GetAbsolutePath(string relativePath) =>
+            Path.Combine(new FileInfo(typeof(
+                TensorFlowScenariosTests).Assembly.Location).Directory.FullName, relativePath);
+
+
+        public class ImageData
+        {
+            [LoadColumn(0)]
+            public string ImagePath;
```
Comment (CESARDELATORRE) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: This input data for the dataset is using FilePaths.
We need to test it with in-memory images. Is there code using and testing in-memory images somewhere else?
